### PR TITLE
Type annotations for torchvision/utils.py

### PR DESCRIPTION
**Pull request by pmeier**
Addresses #2025

### CODE REVIEW

File: torchvision/utils.py [hunk 1]
```diff
@@ -1,10 +1,13 @@
...
 import torch
 import math
 irange = range
 
 
-def make_grid(tensor, nrow=8, padding=2,
-              normalize=False, range=None, scale_each=False, pad_value=0):
+def make_grid(tensor: Union[torch.Tensor, Sequence[torch.Tensor]], nrow: int = 8,
+              padding: int = 2, normalize=False, range: Optional[Tuple[int, int]] = None,
+              scale_each: bool = False, pad_value: int = 0) -> torch.Tensor:
```
Comment (pmeier) [inline torchvision/utils.py]: Can we somehow make the output type dependent on the input type, i.e. if I pass a `torch.DoubleTensor` I should also get back a `torch.DoubleTensor`?

Comment (fmassa) [reply torchvision/utils.py]: I would have expected it to be the case (except for integer tensors), maybe it's an oversight in the code?
I saw that we use the type of the `tensor` in `grid = tensor.new_full(...)`, so it should keep the original `dtype` normally

File: torchvision/utils.py [hunk 2]
```diff
@@ -1,10 +1,13 @@
...
+import io
 import torch
 import math
 irange = range
 
 
-def make_grid(tensor, nrow=8, padding=2,
-              normalize=False, range=None, scale_each=False, pad_value=0):
+def make_grid(tensor: Union[torch.Tensor, Sequence[torch.Tensor]], nrow: int = 8,
+              padding: int = 2, normalize=False, range: Optional[Tuple[int, int]] = None,
```
Comment (fmassa) [inline torchvision/utils.py]: nit:  `normalize: bool = False`, although I would assume that mypy can infer the type annotation from the default argument

File: torchvision/utils.py [hunk 3]
```diff
@@ -88,8 +91,9 @@ def norm_range(t, range):
     return grid
 
 
-def save_image(tensor, fp, nrow=8, padding=2,
-               normalize=False, range=None, scale_each=False, pad_value=0, format=None):
+def save_image(tensor: Union[torch.Tensor, Sequence[torch.Tensor]], fp: Union[str, io.FileIO, io.BytesIO],
+               nrow: int = 8, padding: int = 2, normalize: bool = False, range: Optional[Tuple[int, int]] = None,
+               scale_each: bool = False, pad_value: int = 0, format: Optional[str, io.FileIO] = None) -> None:
```
Comment (fmassa) [inline torchvision/utils.py]: can format be a `FileIO`? From https://github.com/python-pillow/Pillow/blob/349bcd86915958f67b865f5c4c9ef683b7b46756/src/PIL/Image.py#L2094, looks like it calls `.upper()` on it.

File: torchvision/utils.py [hunk 4]
```diff
@@ -88,8 +91,9 @@ def norm_range(t, range):
     return grid
 
 
-def save_image(tensor, fp, nrow=8, padding=2,
-               normalize=False, range=None, scale_each=False, pad_value=0, format=None):
+def save_image(tensor: Union[torch.Tensor, Sequence[torch.Tensor]], fp: Union[str, io.FileIO, io.BytesIO],
```
Comment (fmassa) [inline torchvision/utils.py]: I wonder if there is a canonical way of writing this, as PIL also support `Path` objects https://github.com/python-pillow/Pillow/blob/349bcd86915958f67b865f5c4c9ef683b7b46756/src/PIL/Image.py#L2050

**Comment by fmassa** [comment]
Also, test failures are related
```
TypeError: Optional[t] requires a single type. Got (<class 'str'>, <class '_io.FileIO'>).
```

File: torchvision/utils.py (hunk 1 above)
Comment (pmeier) [reply torchvision/utils.py:10]: I think you misunderstood me. If a function returns the same type as its input, it is my understanding that this is the most generic way to annotate it:

```python
from typing import TypeVar

T = TypeVar("T")


def foo(bar: T) -> T:
    bar.baz()
    return bar
```

Here `bar` could be `torch.FloatTensor`, `torch.ByteTensor` and so on. I don't know if it is possible to limit `T` to `torch.Tensor` and all of its subclasses. Depending on the complexity of the annotations I'm not sure if we even want this.

File: torchvision/utils.py (hunk 3 above)
Comment (pmeier) [reply torchvision/utils.py:96]: Nope, should be `Optional[str]`.

File: torchvision/utils.py (hunk 4 above)
Comment (pmeier) [reply torchvision/utils.py:94]: Maybe I misunderstand you, but we could simply add `pathlib.Path` to the list. If we need to do this more often we can simply create an alias `PILFp = Union[str, io.FileIO, io.BytesIO, pathlib.Path]` and use this within the function signature instead.

**Comment by pmeier** [comment]
Could you link me to the failing test? I can't find it within `circleci` checks.

**Comment by pmeier** [comment]
> Also, I was wondering if we should setup some CI to run mypy in the code (for the files that support type annotation).

Running `mypy torchvision/utils.py` right now results in two classes of errors:

1. `torchvision/utils.py:108: error: Skipping analyzing 'PIL': found module but no type hints or library stubs`
    Since `Pillow` does not provide annotations we either have to provide them ourselves or ignore `PIL` all together. While the first option sounds like a lot of work it really is not that much since we only need provide annotations for the stuff we are using. Here that would be the `Image` class as well as the `fromarray` and `save` method. I've that before in a [project of mine](https://github.com/pmeier/pyimagetest/blob/master/third_party_stubs/PIL/Image.pyi) if you want to have a look.
2. `torchvision/ops/boxes.py:8: error: syntax error in type comment '(Tensor, Tensor, float)'`
    I'm puzzled right now why `mypy` also checks `torchvision/ops/*`, but I think that makes it clear that a CI test as you are proposing will probably not work as intended until we reach a significant annotation coverage. I think including it into CI can help the reviewer that sense that he can see if there any conflicts within the processed file without locally checking out the PR. But if we do this we have to keep in mind that from now on all future PRs will potentially fail the CI checks until we finished annotating.

File: torchvision/utils.py (hunk 1 above)
Comment (fmassa) [reply torchvision/utils.py]: I think annotating it as `torch.Tensor` is enough -- we only have a single type of Tensor now, and `FloatTensor` / `DoubleTensor` are legacy and alias to the same `Tensor` object.

**Comment by fmassa** [comment]
@pmeier 
1 - 
> Since Pillow does not provide annotations we either have to provide them ourselves or ignore PIL all together

Let's ignore PIL annotations (there is a way to tell mypy to ignore annotations from specific modules), and wait until PIL add type annotations to its functions

2 - I think we just need to properly make `mypy` ignore every folder / file except the ones that we want to test.

Additionally, instead of adding the mypy test in `travis.yml`, let's add it instead to CircleCI, similar to what we did in https://github.com/pytorch/vision/pull/2056. For some reason TravisCI has been unstable those past few days, so let's just finish moving away from it.

**Comment by pmeier** [comment]
I'll open a new PR for that and rebase this when the other is merged.

File: torchvision/utils.py (hunk 4 above)
Comment (fmassa) [reply torchvision/utils.py:95]: Yes, we can simply add `pathlib.Path` to the list, but my thinking would be to do whatever PIL does here (as we simply dispatch to PIL anyway).

For example, in a PIL WIP PR they do it as follows: https://github.com/python-pillow/Pillow/pull/2941/files#diff-b49fc26d17e5d3b539a05c7850abe2b0R2025

```
Union[Text, pathlib.Path, BinaryIO]
```

But I don't have any strong opinions here.

Comment (pmeier) [reply torchvision/utils.py:95]: Sure, I agree we should use the same annotations as `PIL`. I wasn't aware that they begun adding them. Thanks for the heads up!

**Comment by pmeier** [comment]
@fmassa I'm confused why you merged this. Does that mean you want me to continue without `mypy` running in CI or shall I hold further plans for type annotations until we figure that out?

**Comment by fmassa** [comment]
@pmeier this PR is very self-contained for now, and I think the next step should be to add CI for type annotations and targeting only this file -- this way we can have signal that our annotations are correct and compliant.

What do you think?

**Pull request merged**